src/events/message/commands/fun/rankup.ts

```diff
@@ -0,0 +1,230 @@
+  --limit=     The max rank to use in calculations ([1-8]kyu)
+*/
+
+enum Mode {
+  Least = "least",
+  Each = "each",
+  Spread = "spread", // default
+}
+const DEFAULTMODE: Mode = Mode.Spread;
+const SPREADWEIGHT: number = 0.6; // used for Spread mode
+
+const LANGS = new Set<string>([
+  "agda",
+  "bf",
+  "c",
+  "cfml",
+  "clojure",
+  "cobol",
+  "coffeescript",
+  "commonlisp",
+  "coq",
+  "cpp",
+  "crystal",
+  "csharp",
+  "d",
+  "dart",
+  "elixir",
+  "elm",
+  "erlang",
+  "factor",
+  "forth",
+  "fortran",
+  "fsharp",
+  "go",
+  "groovy",
+  "haskell",
+  "haxe",
+  "idris",
+  "java",
+  "javascript",
+  "julia",
+  "kotlin",
+  "lean",
+  "lua",
+  "nasm",
+  "nim",
+  "objc",
+  "ocaml",
+  "pascal",
+  "perl",
+  "php",
+  "powershell",
+  "prolog",
+  "purescript",
+  "python",
+  "r",
+  "racket",
+  "raku",
+  "reason",
+  "ruby",
+  "rust",
+  "scala",
+  "shell",
+  "solidarity",
+  "sql",
+  "swift",
+  "typescript",
+  "vb",
+]);
+
+// https://docs.codewars.com/gamification/ranks
+// 3-8 dan added speculatively, based on trend
+const RANKTHRESHOLDS: { [rank: string]: number } = {
+  "7kyu": 20,
+  "6kyu": 76,
+  "5kyu": 229,
+  "4kyu": 643,
+  "3kyu": 1768,
+  "2kyu": 4829,
+  "1kyu": 13147,
+  "1dan": 35759,
+  "2dan": 97225,
+  "3dan": 264324,
+  "4dan": 718593,
+  "5dan": 1953551,
+  "6dan": 5310860,
+  "7dan": 14437910,
+  "8dan": 39250354,
+};
+
+const RANKPOINTS: { [rank: string]: number } = {
+  "8kyu": 2,
+  "7kyu": 3,
+  "6kyu": 8,
+  "5kyu": 21,
+  "4kyu": 55,
+  "3kyu": 149,
+  "2kyu": 404,
+  "1kyu": 1097,
+};
+
+const capitalise = (s: string): string =>
+  s.length ? s[0].toUpperCase() + s.slice(1).toLowerCase() : "";
+
+// Retrieve a users language score, or overall score if language is undefined
+const getUserLangScore = async (user: string, lang?: string): Promise<number | undefined> => {
+  return await fetch("https://www.codewars.com/api/v1/users/" + user)
+    .then((response: any): any => (response.status === 404 ? { success: false } : response.json()))
+    .then((result: any): any => {
+      if (result.success === false) return;
+      return (lang ? result.ranks.languages[lang] : result.ranks.overall)?.score ?? 0;
+    });
+};
+
+export default async (message: Message, args: CommandArg[], opts: Record<string, string>) => {
+  // Helper function
+  const send = (msg: string) => {
+    message.channel.send(msg);
+  };
+  let username: string;
+  if (args.length == 0) {
+    if (!message.member?.displayName) return;
+    username = message.member.displayName;
+  } else {
+    // Input validation
+    const result = z.tuple([word()]).safeParse(args);
+    if (!result.success) {
+      message.reply(USAGE);
+      return;
+    }
+    [username] = result.data;
+  }
+  // Get mode
+  const mode: Mode = (<any>Mode)[capitalise(String(opts.mode).toLowerCase())] ?? DEFAULTMODE;
+
+  // Check that language exists
+  if (opts.language && !LANGS.has(opts.language))
```

Comment:
I would suggest checking correctness of requested language only if the requested language does not exist in data returned for the user(s). This way we avoid necessity of updating the list here immediately after a new language is added. After all, if there is data returned for a language, then its valid regardless of being present on the list.

Reply:
Yeah, that should be good enough. We don't need to maintain the list here.

Reply:
I am not sure I understand how this would help. Either way, the list of languages would need to be here and updated, wouldn't it? Because sometimes a user might query a language which exists, but that they haven't started.

Reply:
I don't remember what I was thinking 6 hours ago. I guess I wasn't considering querying the language the user haven't started yet.

Maintaining the list is not a big deal (I'm sure one of you will update), so I'm OK with this.

Reply:
The other option (how it was handled prior to my last push) is to just show a generic `"${username} has not started training ${language}, or the language ID is wrong"` with a link to the docs maybe. Then there's no need to keep an updated list.

Reply:
Yeah, that's fine too.

```
Kacarott has not started training `brainfuck`, or the language ID is invalid.
See <https://docs.codewars.com/languages/> to find the correct ID.
```